### DISCUSSION TITLE
Hardhat node extremely slow compared to using archive node directly

### DISCUSSION THREAD

**Discussion by LaGregance** [NomicFoundation/hardhat#3766] [Troubleshooting]
Hi,

I build an algorithms that need to perform a lot of `eth_call` to make calculation and then execute a smart contract on Polygon. Using a node from Quicknode, the whole process take around 2 minutes.

For testing purpose I want to run the algorithm under an hardhat node that fork the Polygon network (using the same Quicknode as archive node in hardhat config). I really need hardhat in this case because I want to be able to give me some ERC20 before running tests and also deploy the last version of my contract (I cannot use a live testnet cause my contract had so much interaction with many of DeFi protocols).

Everything is working but when it take 2 minutes when using the archive node directly, it take around 45 minutes under hardhat.

Do you have an idea to increase the hardhat speed or any other way that can solve my problem ?
Thanks.

**Comment by fvictorio** [comment]
Hi @LaGregance. There are two things that make using Hardhat in fork mode significantly slower than using a real node directly:

1. Hardhat uses a javascript implementation of the EVM that is way slower than, say, geth's EVM (or whatever real node you are using). We are working on replacing the execution engine with a rust-based one, but it won't be ready for a couple of months.
2. Besides that, the way an `eth_call` to a real node works is different from how it works with Hardhat in forking mode. When you send an `eth_call` to a node, you just send that request, it gets executed in the node, and you get the response. But the way forking works is that Hardhat will execute the call locally, but it will make a lot of requests to the remote node to fetch things like contracts storage slots, accounts balances, etc. This results in a lot of requests, which are done in a very non-parallelizable way, and so if you are not close to the node, the roundtrips will add up. We do cache this results on disk, so successive executions (if you fork from the same block number) should be faster. But the first execution can be super slow because of this.

Hope that explains it! Let me know if you have more questions.

**Reply by LaGregance** [reply]
Thank you for these interesting details, I have always wondered how it worked internally.
I think a quick workaround to increase the speed is to use websocket instead of http node, but when I do this I get the following error:
`Error HH604: Error running JSON-RPC server: invalid protocol`

Is there a way to use WS instead of HTTP ? (I think about create my own local HTTP proxy that forward request to the node under websocket but seems overkill 😅)

**Reply by fvictorio** [reply]
No, IIRC forking only supports HTTP
